### PR TITLE
feat(datasets) Replace column method in the Query class

### PR DESCRIPTION
**Pull request by fpacifici**
Now that we have query processors we need some easier way to actually tweak the query itself.
This PR introduces a method in the Query class that replaces all (well ... almost all) occurrences of a column with another one.
This is supposed to be used by query processors.

It reflects the behavior of get_all_referenced columns (which does not return all referenced columns but it is missing a few). At this point I believe it is more useful to remain consistent with that method than introducing a different behavior. This will be all fixed as soon as we have a better column abstraction.

test plan:
- unit test added for the corner cases.

### CODE REVIEW

File: snuba/query/query.py [hunk 1]
```diff
@@ -204,3 +219,135 @@ def get_all_referenced_columns(self) -> Sequence[Any]:
...
+                condition[2]
+            ]
+        elif isinstance(condition, (tuple, list)):
+            # nested condition
+            return [
+                self.__replace_col_in_condition(cond, old_column, new_column) for cond in condition
+            ]
+        else:
+            # Don't know what this is
+            return condition
```
Comment (fpacifici) [inline snuba/query/query.py]: I think there are some corner cases where this can happen from the query schema. Though I could not effectively pinpoint a specific one.

File: snuba/query/query.py [hunk 2]
```diff
@@ -204,3 +219,135 @@ def get_all_referenced_columns(self) -> Sequence[Any]:
...
+
+    def __replace_col_in_list(
+        self,
+        expressions: Any,
+        old_column: str,
+        new_column: str,
+    ) -> Sequence[Any]:
+        return [
+            self.__replace_col_in_expression(expr, old_column, new_column)
+            for expr in to_list(expressions)
```
Comment (fpacifici) [inline snuba/query/query.py]: There are several fields that accept either a string or a list of any. The to_list method is used to cope with that across the utility methods in general.

File: snuba/query/query.py [hunk 3]
```diff
@@ -204,3 +219,135 @@ def get_all_referenced_columns(self) -> Sequence[Any]:
...
+                self.get_orderby(),
+                old_column,
+                new_column,
+            ))
+
+        if self.get_aggregations():
+            self.set_aggregations([
+                [
+                    aggr[0],
+                    self.__replace_col_in_list(aggr[1], old_column, new_column),
```
Comment (lynnagara) [inline snuba/query/query.py]: Won't this cause problems since the second item is now coerced into a potentially empty list. e.g. Is ["count()", [], "count"] valid?

File: tests/query/test_col_replacement.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        }
+    ),
+    (
+        # Test simple change in all fields
+        {
+            "selected_columns": ["c1", "c2", "c3"],
+            "conditions": [["c1", "=", "a"]],
+            "arrayjoin": "c1",
+            "groupby": "c1",
+            "aggregations": [["count()", ["c1"], "c1"]],
```
Comment (lynnagara) [inline tests/query/test_col_replacement.py]: This looks like a cross between the 2 syntaxes for aggregation. (The syntax that passes arguments as a list shouldn't have the parens?) Since `count()` doesn't take any argument it's also a weird example

**Review by lynnagara — COMMENTED** [review]
What's the relationship between this and `column_expr`? Is this supposed to completely replace `column_expr` replacements outside of the processor?

**Comment by fpacifici** [comment]
> What's the relationship between this and column_expr? Is this supposed to completely replace column_expr replacements outside of the processor?

Eventually there should be no more replacements outside of the processor. column_expr will be broken into two parts: parsing (what transforms tags[a] into an abstract view) and formatting (what turns that into a Clickhouse query.

So there will be three cases:
- simple dataset specific replacements of column (like replacing group_id with nullIf). That can become a query processor that uses this method
- tags support. I am still debating whether that should be a default datatype you inherit in certain datasets or whether hame them managed by a processor at the beginning.
- escaping rules. That goes into Clickhouse formatting

File: snuba/query/query.py (hunk 3 above)
Comment (fpacifici) [reply snuba/query/query.py]: You are right, this was wrong. I misinterpreted the json schema. This cannot be a list of columns. It can be an expression instead.

File: tests/query/test_col_replacement.py (hunk 4 above)
Comment (fpacifici) [reply tests/query/test_col_replacement.py]: fixed

File: snuba/query/query.py [hunk 5]
```diff
@@ -204,3 +219,138 @@ def get_all_referenced_columns(self) -> Sequence[Any]:
...
+        """
+        Returns a copy of the expression with old_column replaced by new_column.
+        This does not modify anything in place since expressions can be either
+        mutable (lists) or immutable (strings and tuples), so the only way to be
+        consistent in the result is always to return a brand new expression.
+
+        This does not support all possible columns format (like string representations
+        of funciton expresison "count(col1)"). But it is consistent with the behavior
+        of get_all_referenced_columns in that it won't replace something that is not
+        returned by that function and it will replace all the columns reteturned by that
```
Comment (tkaemming) [inline snuba/query/query.py]:
```suggestion
        returned by that function and it will replace all the columns returned by that
```

**Review by tkaemming — APPROVED** [review]
I feel like there are a bunch of edge cases here that are going to be really challenging to anticipate (not issues introduced here — just the nature of the column parsing code), but the general idea makes sense to me, and I think we'll just have to deal with them as they come until these abstractions are cleaner/clearer.

File: snuba/query/query.py [hunk 6]
```diff
@@ -33,6 +34,8 @@
 
 TElement = TypeVar("TElement")
 
+ORDERED_COL = re.compile(r"^-?([a-zA-Z_][a-zA-Z0-9_\.]*)$")
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/query.py]: right, I did not notice it was the same.

File: snuba/query/query.py [hunk 7]
```diff
@@ -204,3 +219,138 @@ def get_all_referenced_columns(self) -> Sequence[Any]:
...
+        ]
+
+    def replace_column(self, old_column: str, new_column: str) -> None:
+        """
+        Replaces a column in all fields of the query. The Query object is mutated in place
+        while the internal fields are replaced.
+
+        This behaves consistently with get_all_referenced_columns (which does not really
+        behave correctly since it is missing a few fields that can contain columns). Will
+        fix both when adding a better column abstraction.
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/query.py]: ok

File: snuba/query/query.py [hunk 8]
```diff
@@ -204,3 +219,138 @@ def get_all_referenced_columns(self) -> Sequence[Any]:
...
+
+        In the current implementation we can only replace a column identified by a string
+        with another column identified by a string. This does not support replacing a
+        column with a complex expression.
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/query.py]: It actually can replace a tag with something else since a tag is a "column represented as a string". Will add a test about that.

**Review by fpacifici — COMMENTED** [review]
yeah, I thought about whether to fix the corner cases with this representation before getting the new abstraction. Finally I did not go that way because, in several cases, the expected behavior itself is not clear, so I suspect it would take longer to nail those behaviors down than to build the new abstraction.

File: snuba/query/query.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/query/query.py]: It's not obvious to me how that'd work because of the regular expression — curious to see if I'm missing something or if it will require changes to support that.

[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/query.py]: I will clarify in the comment. 
This is really not that sophisticated at this level for two reasons:
- this method is always going to be called before any Clickhouse formatting 
- column_expr (which resolves tags) is performed during Clickhosue formatting right now. It will change soon. 

This implies that this method will encounter tags in the form of "tag[aaaa]" which is a string. Potentially this method can be used to replace it with "not_a_tag[not_aaaa]" though it would not be a great idea so it will be discouraged but there is no way we can strictly prevent that without any structure around a tag column. column_expr will be executed later and pick up whatever it sees in the query.

With structured columns things will be better. There are two possible ways to deal with that, I am still thinking over which one is better:
1) resolve tags through a query processor early on.  This turns tags into standard columns (promoted) or into function calls. This is straightforward. But it makes the concept of tag disappear for the following processors. They would not know that the function call came from a tag resolution or to replace a tag with a different tag.
2) parse tag[a] into a specific column type we can manipulate and resolve it right before Clickhouse formatting. This is more complex, it requires dataset to be able to enable/disable some feature during parsing, but it allows query processors to manipulate a higher level concept (the tag) thus being more powerful.
Will resolve this problem in the next days.
